### DISCUSSION TITLE
Flat resource into nested resource?

### DISCUSSION THREAD

**Discussion by state-developer** [okuramasafumi/alba#211] [General]
I have a Rails active record that looks like this:

```
membership_number: 12988,
member_name: Terry Bradshaw,
member_address1: 5555 West Test,
member_city: Miami,
member_state: FL
```

I'm trying to figure out how to transform the object into something a 3rd party API needs (like this):

```
{
  "number": 12988,
  "member": {
    "name": "Terry Bradshaw",
    "address:" {
      "line1:" "5555 West Test",
      "city": "Miami",
      "state": "FL"
     }
  }
}
```

Changing the `membership_number` to `number` is easy enough.  However, I'm getting stuck trying to create the nested part of the JSON from the active record object.  I've tried `one` and various other ways of doing, like creating a `MemberNameResource` and `MemberAddressResource`, but nothing I've tried seems to work.

Here is an example of what I've tried, but it didn't seem to work (simplified):

```
class MemberResource
  include Alba::Resource
  
  attribute :number, :&membership_number
  attributes :member
  
  def member(resource)
    MemberNameResource.new({ resource.slice('member_name') }).to_h
  end
end

class MemberNameResource
  include Alba::Resource
  
  attributes :name, :&member_name
end
```

Any suggestions?

**Comment by state-developer** [comment]
Think I found a solution:

```
class MemberResource
  include Alba::Resource
  
  attribute :number, :&membership_number
  
  attribute :member do |resource|
    MemberNameResource(resource).to_h
  end
end

class MemberNameResource
  include Alba::Resource
  
  attributes :name, :&member_name
end
```

Hope that helps anyone else trying to do the same thing.

**Comment by okuramasafumi** [comment]
Hi @state-developer I think you can use [compose_of](https://api.rubyonrails.org/classes/ActiveRecord/Aggregations/ClassMethods.html#method-i-composed_of) method to aggregate member data, that makes to implement API with Alba much easier.

If we don't want to use `compose_of`, the code will look like this:

```ruby
class MemberAddressResource
  include Alba::Resource

  attribute :lines, &:member_address1
  attribute :city, &:member_city
  attribute :state, &:member_state
end

class MemberInnerResource
  include Alba::Resource

  attribute :name, &:member_name
  attribute :address do |inner|
    MemberAddressResource.new(inner).to_h
  end
end

class MemberResource
  include Alba::Resource

  attribute :number, &:membership_number
  attribute :member do |foo|
    MemberInnerResource.new(foo).to_h
  end
end
```

Alba currently doesn't have an API to define nested hash easily. If we need it so much I think we should implement it.

**Reply by state-developer** [reply]
Yes.  This is what I ended up doing.  Thanks!
